pennylane/labs/lie/structure_constants.py

```diff
@@ -0,0 +1,80 @@
+
+
+def structure_constants_dense(g: TensorLike, is_orthogonal: bool = False) -> TensorLike:
+    r"""
+    Compute the structure constants that make up the adjoint representation of a Lie algebra.
+
+    This function computes the structure constants of a Lie algebra provided by their dense matrix representation,
+    obtained from, e.g., :func:`~lie_closure_dense`.
+    This is sometimes more efficient than using the sparse Pauli representations of :class:`~PauliWord` and
+    `~PauliSentence` that are employed in :func:`~structure_constants`, e.g., when there are few generators
+    that are sums of many Paulis.
+
+    .. seealso:: For details on the mathematical definitions, see :func:`~structure_constants` and the section "Lie algebra basics" in our `g-sim demo <https://pennylane.ai/qml/demos/tutorial_liesim/#lie-algebra-basics>`__.
+
+    Args:
+        g (np.array): The (dynamical) Lie algebra provided as dense matrices, as generated from :func:`pennylane.labs.lie.lie_closure_dense`.
+            ``g`` should have shape ``(d, 2**n, 2**n)`` where ``d`` is the dimension of the algebra and ``n`` is the number of qubits.
+        is_orthogonal (bool): Whether or not the matrices in ``g`` are orthogonal with respect to the Hilbert-Schmidt inner product on
+            (skew-)Hermitian matrices. If the inputs are orthogonal, it is recommended to set ``is_orthogonal`` to ``True`` to reduce
+            computational cost. Defaults to ``False``.
+
+    Returns:
+        TensorLike: The adjoint representation of shape ``(d, d, d)``, corresponding to indices ``(gamma, alpha, beta)``.
+
+    **Example**
+
+    Let us generate the DLA of the transverse field Ising model using :func:`~lie_closure_dense`.
+
+    >>> n = 4
+    >>> gens = [X(i) @ X(i+1) + Y(i) @ Y(i+1) + Z(i) @ Z(i+1) for i in range(n-1)]
+    >>> g = lie_closure_dense(gens)
+    >>> g.shape
+    (12, 16, 16)
+
+    The DLA is represented by a collection of twelve :math:`2^4 \times 2^4` matrices.
+    Hence, the dimension of the DLA is :math:`d = 12` and the structure constants have shape ``(12, 12, 12)``.
+
+    >>> adj = structure_constants_dense(g)
+    >>> adj.shape
+    (12, 12, 12)
+
+    """
+    dimg, chi, _ = g.shape
+    assert g.shape[2] == g.shape[1]
+
+    # compute all commutators by computing all products first according to "aij,bjk->abik"
+    prod = np.moveaxis(np.tensordot(g, g, axes=[[2], [1]]), 1, 2)
+    all_coms = np.reshape(prod - prod.transpose((1, 0, 2, 3)), (-1, chi, chi))
+
+    # project commutators on basis of g
+    # vectorized computation to obtain coefficients in decomposition:
+    # v = ∑ (tr(v @ e_j) / ||e_j||^2) * e_j
+    if is_orthogonal:
+        # Compute norms of orthogonal entries
+        norms = np.einsum("aij,aji->a", g, g).real
+        gram_inv = np.diag(1 / norms)
+    else:
+        # Compute the full inverse Gram matrix of the entries.
+        gram_inv = np.linalg.pinv(np.tensordot(g, g, axes=[[1, 2], [2, 1]])).real
```

Comment:
I think I just now realize that this is not just normalizing but actually orthonormalizing the basis, right? So shouldnt we just orthonormalize g at the start and then continue with that?

Reply:
Not sure actually / confusion to the highest form :D thinking out loud here: I thought this would normalize a basis:

```
g = np.random.rand(4, 8, 8)
g = np.einsum("aij,aik->ajk", g, g) # make Hermitian

assert np.allclose(g, np.transpose(g, (0, 2, 1))) # assert Hermitian

normalizer = np.linalg.pinv(np.tensordot(g, g, axes=[[1, 2], [2, 1]])).real

g_norm = np.tensordot(normalizer, g, axes=([1], [0]))

norms = np.tensordot(g_norm, g_norm, axes=([1, 2], [2, 1])).real
np.allclose(norms, np.eye(4))
```

Reply:
I think you need a square root in there, don't you? :thinking: 

I would like to keep the input untouched if we can, so I'd go for the existing approach that adapts the inner products to the non-orthonormal case, using the Gram matrix, rather than changing the input matrices and computing the adjoint representation of the modified input.
If we want a function that orthonormalizes and then computes the adjoint rep, we should just chain `_hermitian_basis` and `structure_constants` :)

Reply:
Indeed, this works:
```python
g = np.random.rand(4, 8, 8)
g = np.einsum("aij,aik->ajk", g, g) # make Hermitian

assert np.allclose(g, np.transpose(g, (0, 2, 1))) # assert Hermitian

normalizer = scipy.linalg.sqrtm(np.linalg.pinv(np.tensordot(g, g, axes=[[1, 2], [2, 1]])).real)

g_norm = np.tensordot(normalizer, g, axes=([1], [0]))

norms = np.tensordot(g_norm, g_norm, axes=([1, 2], [2, 1])).real
np.allclose(norms, np.eye(4)) # would have expected to be True
# True
```

Reply:
> I would like to keep the input untouched if we can, so I'd go for the existing approach that adapts the inner products to the non-orthonormal case, using the Gram matrix, rather than changing the input matrices and computing the adjoint representation of the modified input.

Would that not result in the same adjoint representation 🤔 ? (if it is really just about normalization, I agree that for non-orthogonal elements it would be different, but does the adjoint representation even make sense for non-orthogonal bases?)

Reply:
Mhmhm. Let me fiddle around with this a bit. I'd think that the representation may differ because it is a quadratic-vs-linear equation underlying its definition:
Say I have `g=span{A, B, C}` orthonormal, and `2C = [A, B]`, so that f^c_{ab} = 2.
Then changing to an orthogonal, but unnormalized version, say `g'=span{A, B'=2B, C}`, we get
`4C = [A, B']`, so that `f'^c_{ab} = 4`, even when including the `gram_inv` step from the code, because we are renormalizing by the norm of `C`, which was untouched and hence is 1.